vscode/src/autoedits/renderer.ts

```diff
@@ -1,202 +1,536 @@
-        }
-
-        editor.setDecorations(
-            strikeThroughDecorationType,
-            removedLines.map(line => ({
-                range: new vscode.Range(line.line, 0, line.line, document.lineAt(line.line).text.length),
-            }))
-        )
-        editor.setDecorations(
-            suggesterType,
-            addedLines.map(line => ({
-                range: new vscode.Range(line.line, 0, line.line, document.lineAt(line.line).text.length),
-                renderOptions: {
-                    after: {
-                        contentText: line.text,
-                    },
-                },
-            }))
+        this.logDiff(
+            options.document.uri,
+            options.currentFileText,
+            options.prediction,
+            options.predictedFileText
         )
+        this.activeEdit.renderer.renderDecorations({
+            document: options.document,
+            currentFileText: options.currentFileText,
+            predictedFileText: options.predictedFileText,
+        })
         await vscode.commands.executeCommand('setContext', 'cody.supersuggest.active', true)
     }
 
-    async acceptProposedChange(): Promise<void> {
-        if (this.activeProposedChange === null) {
-            return
+    private async dismissEdit(): Promise<void> {
+        const renderer = this.activeEdit?.renderer
+        if (renderer) {
+            renderer.clearDecorations()
+            renderer.dispose()
         }
+        this.activeEdit = null
+        await vscode.commands.executeCommand('setContext', 'cody.supersuggest.active', false)
+    }
+
+    private async acceptEdit(): Promise<void> {
         const editor = vscode.window.activeTextEditor
-        if (!editor) {
-            await this.dismissProposedChange()
+        if (!this.activeEdit || !editor || editor.document.uri.toString() !== this.activeEdit.uri) {
+            await this.dismissEdit()
             return
         }
-        const currentActiveChange = this.activeProposedChange
         await editor.edit(editBuilder => {
-            editBuilder.replace(currentActiveChange.range, currentActiveChange.newText)
+            editBuilder.replace(this.activeEdit!.range, this.activeEdit!.prediction)
         })
-        await this.dismissProposedChange()
+        await this.dismissEdit()
     }
 
-    async dismissProposedChange(): Promise<void> {
-        this.activeProposedChange = null
-        await vscode.commands.executeCommand('setContext', 'cody.supersuggest.active', false)
-        const editor = vscode.window.activeTextEditor
-        if (!editor) {
+    private async onDidChangeTextDocument(event: vscode.TextDocumentChangeEvent): Promise<void> {
+        // Only dismiss if we have an active suggestion and the changed document matches
+        // else, we will falsely discard the suggestion on unrelated changes such as changes in output panel.
+        if (!this.activeEdit || event.document.uri.toString() !== this.activeEdit.uri) {
```

Comment:
You could simplify slightly this way:

```suggestion
        if (event.document.uri.toString() !== this.activeEdit?.uri) {
```

Reply:
done